### PR TITLE
generate a static tor onion address persistent unique to the node id

### PR DESCRIPTION
**Pull request by Saibato**
We now can generate a V3 tor service address from a blob 
 
 --addr=**statictor**:127.0.0.1:9051[/**torblob**=[blob]]

Or generate a unique static tor address bound to our node id /pub key     

 --addr=**statictor**:127.0.0.1:9051

TODO:

- [x] Documentation

- [x] Write some simple py-tests 

EDIT: With this you are able to keep your public known or group known tor address static
          and independent of torrc file or the tor instance ( if multiple ), in case you have to  
          move your node to an other PC or system/phone, as long as there is a tor service
          your node will be reachable even if ip, NAT, firewall have changed by that static V3 address.

EDIT: related issues #2476  #3085

EDIT_2:

With this PR we can now also select the extern tor port we assign our local port to 

 --addr=**statictor**:127.0.0.1:9051[/**torblob**=[blob]][/**torport**=EXTERNPORT]

Or generate a unique static tor address bound to our node id /pub key     

 --addr=**statictor**:127.0.0.1:9051[/**torport**=EXTERNPORT]

### CODE REVIEW

**Comment by Saibato** [comment]
@cdecker rebased and ready to rumble ... :+1:

File: common/wireaddr.h [hunk 1]
```diff
@@ -120,8 +123,13 @@ struct wireaddr_internal {
...
 		/* ADDR_INTERNAL_ALLPROTO */
 		u16 port;
-		/* ADDR_INTERNAL_AUTOTOR */
-		struct wireaddr torservice;
+		/* ADDR_INTERNAL_AUTOTOR
+		 * ADDR_INTERNAL_STATICTOR */
+		struct torservice {
+			struct wireaddr torservice_address;
+			u16 port;
+			char blob[TOR_V3_BLOBLEN + 1];
```
Comment (rustyrussell) [inline common/wireaddr.h]: u8 blob, instead of char?  And I'm not sure if the +1 is needed...

File: common/wireaddr.c [hunk 2]
```diff
@@ -453,15 +472,69 @@ bool parse_wireaddr_internal(const char *arg, struct wireaddr_internal *addr,
...
 		return parse_wireaddr(arg + strlen("autotor:"),
-				      &addr->u.torservice, 9051,
+				      &addr->u.torservice.torservice_address, 9051,
+				      dns_ok ? NULL : &needed_dns,
+				      err_msg);
+	}
+
+	/* 'statictor:' is a special prefix meaning talk to Tor to create
+	 * an onion address. */
+	if (strstarts(arg, "statictor:")) {
```
Comment (rustyrussell) [inline common/wireaddr.c]: Let's test this once, maybe, not 4 times?

Comment (rustyrussell) [reply common/wireaddr.c]: Actually, let's use tal_strspit() here:

```
if (strstarts(arg, "statictor:)) {
    /* Format is separated by colons. */
    char **parts = tal_strsplit(tmpctx, arg, ":", STR_EMPTY_OK);
    for (size_t i = 1; i < tal_count(parts)-1; i++) { 
        if (streq(parts[i], "torport")) {
            char **endp;
            addr->u.torservice.port = strtol(parts[i], &endp, 10);
            if (addr->u.torservice.port == 0 || endp != '\0') {
                *err_msg = "Bad :torport: number";
                return false;
            }
        } else if (streq(pars[i], "torblob")) {
		...
```

File: connectd/connectd.c [hunk 3]
```diff
@@ -967,6 +976,23 @@ static void finalize_announcable(struct wireaddr **announcable)
...
+		if (bindings[i].itype != ADDR_INTERNAL_WIREADDR)
+			continue;
+		if (bindings[i].u.wireaddr.type != ADDR_TYPE_IPV4
+		    && bindings[i].u.wireaddr.type != ADDR_TYPE_IPV6)
+			continue;
+		return &bindings[i].u.wireaddr;
+	}
+	status_failed(STATUS_FAIL_INTERNAL_ERROR,
+		      "No local address found to tell Tor to connect to");
+}
```
Comment (rustyrussell) [inline connectd/connectd.c]: Make this non static and delete the one in tor_autoservice.c?  Having two is just going to confuse me!

File: connectd/connectd.c [hunk 4]
```diff
@@ -1097,28 +1128,71 @@ static struct wireaddr_internal *setup_listeners(const tal_t *ctx,
...
+				if (sodium_mlock(&random, sizeof(random)) != 0)
+						status_failed(STATUS_FAIL_INTERNAL_ERROR,
+									"Could not lock the random prf key memory.");
+				randombytes_buf((void * const)&random, 32);
+				/* generate static tor node address, take first 32 bytes from secret of node_id plus 32 random bytes from sodiom */
+				struct sha256 sha;
+				sha256(&sha, hsm_do_ecdh(tmpctx, &pb), 32);
+				/* even if it's a partly secret, tor shall see only the sha */
+				blob = tal_fmt(tmpctx, "%.32s%.32s", tal_hexstr(tmpctx, &sha, 32), tal_hexstr(tmpctx, &random, 32));
+			}
```
Comment (rustyrussell) [inline connectd/connectd.c]: sodium_munlock(&random) is needed at the end.  Which also zeroes.

Seems like random can be a local var here, too?  As can pb...

File: connectd/connectd.c [hunk 5]
```diff
@@ -1097,28 +1128,71 @@ static struct wireaddr_internal *setup_listeners(const tal_t *ctx,
...
+						0);
+				continue;
+		};
+		add_announcable(announcable,
+				tor_fixed_service(tmpctx,
+						&proposed_wireaddr[i],
+						tor_password,
+						blob,
+						find_local_address(binding),
+						0));
```
Comment (rustyrussell) [inline connectd/connectd.c]: Let's use a temporary here:

```
toraddr = tor_fixed_service(....);
if (proposed_listen_announce[i] & ADDR_ANNOUNCE)
        add_announcable(announcable, toraddr);
```

File: connectd/tor_autoservice.c [hunk 6]
```diff
@@ -133,6 +152,54 @@ static struct wireaddr *make_onion(const tal_t *ctx,
...
+
+	blob64 = b64_encode(tmpctx,(char *) blob, 64);
+
+	tor_send_cmd(rbuf,
+			  tal_fmt(tmpctx, "ADD_ONION ED25519-V3:%s Port=%d,%s Flags=DiscardPK",
+					blob64, port, fmt_wireaddr(tmpctx, local)));
+
+	while ((line = tor_response_line_wfail(rbuf)) != NULL) {
+		const char *name;
+		if (line && strstarts(line, "Onion address collision"))
```
Comment (rustyrussell) [inline connectd/tor_autoservice.c]: Don't need to check if `line` is NULL, since loop would have exited.

File: connectd/tor_autoservice.c [hunk 7]
```diff
@@ -133,6 +152,54 @@ static struct wireaddr *make_onion(const tal_t *ctx,
...
+			status_failed(STATUS_FAIL_INTERNAL_ERROR,
+				      "Tor address in use\n");
```
Comment (rustyrussell) [inline connectd/tor_autoservice.c]: No `\n` needed on status_ args.

File: connectd/tor_autoservice.c [hunk 8]
```diff
@@ -133,6 +152,54 @@ static struct wireaddr *make_onion(const tal_t *ctx,
...
+		#else
+		status_info("Static Tor service onion address: \"%s:%d,%s\" bound to external port %d ",
+						name, DEFAULT_PORT ,fmt_wireaddr(tmpctx, local), port);
+		#endif
+		discard_remaining_response(rbuf);
+		return onion;
+	}
+	return NULL;
+	status_failed(STATUS_FAIL_INTERNAL_ERROR,
+		      "Tor didn't give us a ServiceID");
```
Comment (rustyrussell) [inline connectd/tor_autoservice.c]: Unreachable code?

File: connectd/tor_autoservice.c [hunk 9]
```diff
@@ -263,3 +333,45 @@ struct wireaddr *tor_autoservice(const tal_t *ctx,
...
+	if (connect(fd, ai_tor->ai_addr, ai_tor->ai_addrlen) != 0)
+		err(1, "Connecting stream socket to Tor service");
+
+	buffer = tal_arr(tmpctx, char, rbuf_good_size(fd));
+	rbuf_init(&rbuf, fd, buffer, tal_count(buffer), buf_resize);
+
+	negotiate_auth(&rbuf, tor_password);
+
+	onion = make_fixed_onion(ctx, &rbuf, laddr, blob, tor_serviceaddr->u.torservice.port);
+	memset((void *)blob, 0, sizeof(*blob));
```
Comment (rustyrussell) [inline connectd/tor_autoservice.c]: Don't destroy this here, it's const so we promised we wouldn't!  If we really want to do this, we need to do it at a higher level.

File: doc/lightningd-config.5 [hunk 10]
```diff
@@ -304,6 +305,20 @@ and this will be used to configure a Tor hidden service for port
...
 
+If the argument begins with 'statictor:' then it is followed by the
+IPv4 or IPv6 address of the Tor control port (default port 9051),
+and this will be used to configure a static Tor hidden service for port
+9735.  The Tor hidden service will be configured to point to the
+first IPv4 or IPv6 address we bind to and is by default unique to
+your nodes id. You can add the text ':torblob:' followed by up to
+64 Bytes of text to generate from this text a v3 onion service
+address unique to the first 32 bytes of this text. The later 32 bytes
+will deteremine mainly your Tor onions inner secret but will not
```
Comment (rustyrussell) [inline doc/lightningd-config.5]: determine

File: doc/lightningd-config.5 [hunk 11]
```diff
@@ -304,6 +305,20 @@ and this will be used to configure a Tor hidden service for port
...
+change the text of the base32 encoded tor address.
+Those remaining 32 bytes should and can be randomly selected.
+at startup. The text of the onion address is detremined by
+the firt 32 bytes of the blob.
```
Comment (rustyrussell) [inline doc/lightningd-config.5]: determined by the first

File: common/base64.c [hunk 12]
```diff
@@ -0,0 +1,34 @@
+#include <common/base64.h>
+#include <sodium.h>
+#include <sodium/utils.h>
+
+/* Decode/encode from/to base64, base64 helper functions.
+ * We import base64 from libsodium to generate tor V3 ED25519-V3 onions from blobs
+*/
+
+char *b64_encode(const tal_t *ctx, char *data, size_t len)
```
[the remark this answers is not in the document]
Comment (Saibato) [reply common/base64.c]: done

File: common/base64.c [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+{
+	char *str = tal_arrz(ctx, char, sodium_base64_encoded_len(len, sodium_base64_VARIANT_ORIGINAL) + 1);
```
[the remark this answers is not in the document]
Comment (Saibato) [reply common/base64.c]: done

File: common/base64.c [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+
+	str = sodium_bin2base64(str,  tal_count(str), (const unsigned char *)data,
+								len, sodium_base64_VARIANT_ORIGINAL);
+	return str;
+}
+
+char *b64_decode(const tal_t *ctx, char *str, size_t len)
```
[the remark this answers is not in the document]
Comment (Saibato) [reply common/base64.c]: done

File: common/base64.c [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+{
+	char *ret = tal_arrz(ctx, char, strlen(str) + 1);
```
[the remark this answers is not in the document]
Comment (Saibato) [reply common/base64.c]: done

File: common/base64.c [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+	const char *b64_end;
+
+	if (!sodium_base642bin((unsigned char * const )ret,
+				tal_count(ret),
+				(const char * const)str,
+				len,
+				NULL,
+				NULL,
+				&b64_end,
```
[the remark this answers is not in the document]
Comment (Saibato) [reply common/base64.c]: done

File: common/wireaddr.h (hunk 1 above)
Comment (Saibato) [reply common/wireaddr.h]: done; I kept the + 1 for now ...

File: common/wireaddr.c (hunk 2 above)
Comment (Saibato) [reply common/wireaddr.c]: done, now with tal_strsplit, incl. ipv6 ::  handling. TODO: needs some test

File: connectd/connectd.c (hunk 3 above)
Comment (Saibato) [reply connectd/connectd.c]: Now non static in tor_autoservice.c ( fewer changes )

File: connectd/connectd.c (hunk 4 above)
Comment (Saibato) [reply connectd/connectd.c]: done. changed.

File: connectd/connectd.c (hunk 5 above)
Comment (Saibato) [reply connectd/connectd.c]: now temp, done

File: connectd/tor_autoservice.c (hunk 6 above)
Comment (Saibato) [reply connectd/tor_autoservice.c]: ;-)

File: connectd/tor_autoservice.c (hunk 7 above)
Comment (Saibato) [reply connectd/tor_autoservice.c]: done

File: connectd/tor_autoservice.c (hunk 8 above)
Comment (Saibato) [reply connectd/tor_autoservice.c]: ooops, removed

File: connectd/tor_autoservice.c (hunk 9 above)
Comment (Saibato) [reply connectd/tor_autoservice.c]: done, clear data moved to connectd.c

File: doc/lightningd-config.5 (hunk 10 above)
Comment (Saibato) [reply doc/lightningd-config.5]: :-)

File: doc/lightningd-config.5 (hunk 11 above)
Comment (Saibato) [reply doc/lightningd-config.5]: ...

**Comment by Saibato** [comment]
@rustyrussell  thanks for review :+1:   , I changed most lines like you suggested.

If Travis gets happy, you might want to review again.
Its rebased but not sqashed yet. 

Documentation and tests are still rudimentary, sigh ... some work to do.

**Comment by Saibato** [comment]
Tyi @rustyrussell  with your commit https://github.com/rustyrussell/lightning/commit/1e244d5145ebb024c03c023905e89e143c0bc2a2 it passed the Travis lottery. :game_die: :game_die: :game_die: 
so tACK 1e244d5145ebb024c03c023905e89e143c0bc2a2

File: doc/lightningd-config.5.md [hunk 17]
```diff
@@ -261,6 +262,23 @@ Set an IP address (v4 or v6) or automatic Tor address to listen on and
...
+    64 Bytes of text to generate from this text a v3 onion service
+    address unique to the first 32 bytes of this text. The later 32 bytes
+    will determine mainly your Tor onions inner secret but will not
+    change the text of the base32 encoded tor address.
+    Those remaining 32 bytes should and can but must not be randomly selected.
+    at startup. The text of the onion address is determined by
+    the first 32 bytes of the blob.
+    You can also use an postfix ':torport:TORPORT' to select the external
+    tor binding. The result is that over tor your node is accessible by a port
+    defined by you and possible different from your local node port assignment
```
[the remark this answers is not in the document]
Comment (Saibato) [reply doc/lightningd-config.5.md]: Yeah, fewer words say more ...

File: connectd/connectd.c [hunk 18]
```diff
@@ -1109,25 +1124,62 @@ static struct wireaddr_internal *setup_listeners(const tal_t *ctx,
...
+			if (strstr((char *)proposed_wireaddr[i].u.torservice.blob, STATIC_TOR_MAGIC_STRING)) {
+				if (sodium_mlock(&random, sizeof(random)) != 0)
+						status_failed(STATUS_FAIL_INTERNAL_ERROR,
+									"Could not lock the random prf key memory.");
+				randombytes_buf((void * const)&random, 32);
+				/* generate static tor node address, take first 32 bytes from secret of node_id plus 32 random bytes from sodiom */
+				struct sha256 sha;
+				sha256(&sha, hsm_do_ecdh(tmpctx, &pb), 32);
+				/* even if it's a partly secret, tor shall see only the sha */
+				blob = tal_fmt(tmpctx, "%.32s%.32s", tal_hexstr(tmpctx, &sha, 32), tal_hexstr(tmpctx, &random, 32));
```
[the remark this answers is not in the document]
Comment (Saibato) [reply connectd/connectd.c]: Might all be a bit over addressed here, my first version used the priv ;-)
Partly addressed

File: tests/test_gossip.py [hunk 19]
```diff
@@ -1587,3 +1587,65 @@ def test_gossip_ratelimit(node_factory):
...
+            return
+        assert l1.daemon.is_in_log('9735,127.0.0.1:')
+        assert l2.daemon.is_in_log('x2y4zvh4fn5q3eouuh7nxnc7zeawrqoutljrup2xjtiyxgx3emgkemad.onion:9735,127.0.0.1:12345')
+
+        return
+    assert l1.daemon.is_in_log('9735,127.0.0.1:')
+    assert l2.daemon.is_in_log('x2y4zvh4fn5q3eouuh7nxnc7zeawrqoutljrup2xjtiyxgx3emgkemad.onion:9735,127.0.0.1:12345')
+
+
+@unittest.skipIf(not DEVELOPER, "needs a running Tor service instance at port 9151 or 9051")
```
[the remark this answers is not in the document]
Comment (Saibato) [reply tests/test_gossip.py]: We use now check_socket to scan for Tor :-)

File: common/base64.c [hunk 20]
```diff
@@ -0,0 +1,37 @@
...
+	str = sodium_bin2base64(str,  tal_count(str), data,
+								len, sodium_base64_VARIANT_ORIGINAL);
+	return str;
+}
+
+u8 *b64_decode(const tal_t *ctx, const char *str, size_t len)
+{
+	size_t bin_len = len + 1;
+
+	u8 *ret = tal_arrz(ctx, u8, bin_len);
```
[the remark this answers is not in the document]
Comment (Saibato) [reply common/base64.c]: done, arr

File: common/wireaddr.c [hunk 21]
```diff
@@ -13,10 +14,13 @@
 #include <fcntl.h>
 #include <netdb.h>
 #include <netinet/in.h>
+#include <sodium/randombytes.h>
 #include <sys/types.h>
 #include <unistd.h>
 #include <wire/wire.h>
 
+
+
```
[the remark this answers is not in the document]
Comment (Saibato) [reply common/wireaddr.c]: done

File: common/wireaddr.h [hunk 22]
```diff
@@ -120,8 +123,13 @@ struct wireaddr_internal {
 		struct wireaddr wireaddr;
 		/* ADDR_INTERNAL_ALLPROTO */
 		u16 port;
-		/* ADDR_INTERNAL_AUTOTOR */
-		struct wireaddr torservice;
+		/* ADDR_INTERNAL_AUTOTOR
+		 * ADDR_INTERNAL_STATICTOR */
+		struct torservice {
+			struct wireaddr torservice_address;
```
[the remark this answers is not in the document]
Comment (Saibato) [reply common/wireaddr.h]: done

File: common/wireaddr.c [hunk 23]
```diff
@@ -80,7 +84,14 @@ void towire_wireaddr_internal(u8 **pptr, const struct wireaddr_internal *addr)
...
 	case ADDR_INTERNAL_AUTOTOR:
-		towire_wireaddr(pptr, &addr->u.torservice);
+		towire_wireaddr(pptr, &addr->u.torservice.torservice_address);
+		towire_u16(pptr, addr->u.torservice.port);
+		return;
+	case ADDR_INTERNAL_STATICTOR:
+		towire_wireaddr(pptr, &addr->u.torservice.torservice_address);
+		towire_u8_array(pptr, (const u8 *)addr->u.torservice.blob,
+				sizeof(addr->u.torservice.blob));
+		towire_u16(pptr, addr->u.torservice.port);
```
[the remark this answers is not in the document]
Comment (Saibato) [reply common/wireaddr.c]: hmm... I am so lazy must I ?  more code more bugs ? we need this only  once.

File: common/wireaddr.c [hunk 24]
```diff
@@ -113,7 +124,15 @@ bool fromwire_wireaddr_internal(const u8 **cursor, size_t *max,
...
-		return fromwire_wireaddr(cursor, max, &addr->u.torservice);
+		fromwire_wireaddr(cursor, max, &addr->u.torservice.torservice_address);
+		addr->u.torservice.port = fromwire_u16(cursor, max);
+		return *cursor != NULL;
+	case ADDR_INTERNAL_STATICTOR:
+		fromwire_wireaddr(cursor, max, &addr->u.torservice.torservice_address);
+		fromwire_u8_array(cursor, max, (u8 *)addr->u.torservice.blob,
+				 sizeof(addr->u.torservice.blob));
+		addr->u.torservice.port = fromwire_u16(cursor, max);
+		return *cursor != NULL;
```
[the remark this answers is not in the document]
Comment (Saibato) [reply common/wireaddr.c]: ;-)

File: connectd/tor_autoservice.c [hunk 25]
```diff
@@ -122,17 +142,62 @@ static struct wireaddr *make_onion(const tal_t *ctx,
...
+{
+	char *line;
+	struct wireaddr *onion;
+	char *blob64;
+
+	blob64 = b64_encode(tmpctx,(const u8 *) blob, 64);
+
+	tor_send_cmd(rbuf,
+			  tal_fmt(tmpctx, "ADD_ONION ED25519-V3:%s Port=%d,%s Flags=DiscardPK",
+					blob64, port, fmt_wireaddr(tmpctx, local)));
```
[the remark this answers is not in the document]
Comment (Saibato) [reply connectd/tor_autoservice.c]: done

File: connectd/tor_autoservice.c [hunk 26]
```diff
@@ -45,6 +42,27 @@ static void tor_send_cmd(struct rbuf *rbuf, const char *cmd)
...
+			status_failed(STATUS_FAIL_INTERNAL_ERROR,
+				      "Tor returned '%s'", line);
+
+		/* Last line */
+		if (strstarts(line, "250 ") || strstarts(line, "550 "))
+			break;
+
+		return line + 4;
+	}
+	return line + 4;
```
[the remark this answers is not in the document]
Comment (Saibato) [reply connectd/tor_autoservice.c]: hmm... addressed that now  by check for NULL

File: connectd/tor_autoservice.c [hunk 27]
```diff
@@ -194,6 +259,9 @@ static void negotiate_auth(struct rbuf *rbuf, const char *tor_password)
 					     tal_hexstr(tmpctx,
 							contents,
 							tal_count(contents)-1)));
+			//status_info("coo:  %s",tal_hexstr(tmpctx,
+			//				contents,
+			//				tal_count(contents)-1));
```
[the remark this answers is not in the document]
Comment (Saibato) [reply connectd/tor_autoservice.c]: done

File: connectd/tor_autoservice.c [hunk 28]
```diff
@@ -263,3 +331,45 @@ struct wireaddr *tor_autoservice(const tal_t *ctx,
...
+	if (connect(fd, ai_tor->ai_addr, ai_tor->ai_addrlen) != 0)
+		err(1, "Connecting stream socket to Tor service");
+
+	buffer = tal_arr(tmpctx, char, rbuf_good_size(fd));
+	rbuf_init(&rbuf, fd, buffer, tal_count(buffer), buf_resize);
+
+	negotiate_auth(&rbuf, tor_password);
+
+	onion = make_fixed_onion(ctx, &rbuf, laddr, blob, tor_serviceaddr->u.torservice.port);
+	memset((void *)blob, 0, sizeof(*blob));
```
[the remark this answers is not in the document]
Comment (Saibato) [reply connectd/tor_autoservice.c]: annnnnnnnd gone ...

File: common/base64.c [hunk 29]
```diff
@@ -0,0 +1,34 @@
...
+				tal_count(ret),
+				(const char * const)str,
+				len,
+				NULL,
+				NULL,
+				&b64_end,
+				sodium_base64_VARIANT_ORIGINAL))
+			return tal_free(ret);
+
+	return ret;
```
[the remark this answers is not in the document]
Comment (Saibato) [reply common/base64.c]: done

**Comment by Saibato** [comment]
A bit of side show, but almost all PR not only this, fail the Travis lottery on first or second push with random errors unrelated to the changes.
@rustyrussell @cdecker  #3286 solved all ? Travis failed https://travis-ci.org/ElementsProject/lightning/builds/616642255?utm_source=github_status&utm_medium=notification
I will now force push the same PR again and again :woman_facepalming: until I win the Travis check lottery :game_die: :game_die:  :game_die:  take two won ...

**Comment by cdecker** [comment]
> A bit of side show, but almost all PR not only this, fail the Travis lottery on first or second push with random errors unrelated to the changes.
> @rustyrussell @cdecker #3286 solved all ? Travis failed https://travis-ci.org/ElementsProject/lightning/builds/616642255?utm_source=github_status&utm_medium=notification
> I will now force push the same PR again and again  until I win the Travis check lottery    take two won ...

I'm restarting failed tests and keeping track of which tests were flaky, so you don't need to push just to get Travis to like your PR :wink:

**Comment by cdecker** [comment]
Excellent changes @Saibato, sorry for the delay in reviewing it :+1:

ACK https://github.com/ElementsProject/lightning/pull/3155/commits/14882ec1fe261f6d67fc1da937d17c56c76232dc

**Comment by jb55** [comment]
very cool, thanks @Saibato !

**Pull request merged**